djautotask/sync.py

```diff
@@ -0,0 +1,219 @@
+import logging
+
+from suds.client import Client
+from atws import connect, Query
+
+from django.conf import settings
+from django.db import transaction, IntegrityError
+from django.utils import timezone
+
+
+from djautotask import models
+
+logger = logging.getLogger(__name__)
+
+
+class InvalidObjectException(Exception):
+    """
+    If for any reason an object can't be created (for example, it references
+    an unknown foreign object, or is missing a required field), raise this
+    so that the synchronizer can catch it and continue with other records.
+    """
+    pass
+
+
+def log_sync_job(f):
+    def wrapper(*args, **kwargs):
+        sync_instance = args[0]
+        created_count = updated_count = deleted_count = 0
+        sync_job = models.SyncJob()
+        sync_job.start_time = timezone.now()
+        if sync_instance.full:
+            sync_job.sync_type = 'full'
+        else:
+            sync_job.sync_type = 'partial'
+
+        try:
+            created_count, updated_count, deleted_count = f(*args, **kwargs)
+            sync_job.success = True
+        except Exception as e:
+            sync_job.message = str(e.args[0])
+            sync_job.success = False
+            raise
+        finally:
+            sync_job.end_time = timezone.now()
+            sync_job.entity_name = sync_instance.model_class.__name__
+            sync_job.added = created_count
+            sync_job.updated = updated_count
+            sync_job.deleted = deleted_count
+            sync_job.save()
+
+        return created_count, updated_count, deleted_count
+    return wrapper
+
+
+class SyncResults:
+    """Track results of a sync job."""
+    def __init__(self):
+        self.created_count = 0
+        self.updated_count = 0
+        self.deleted_count = 0
+        self.synced_ids = set()
+
+
+class Synchronizer:
+    lookup_key = 'id'
+
+    def __init__(self, full=False, *args, **kwargs):
+        self.full = full
+
+        self.at_api_object = self.init_api_connection()
```

Comment:
Something to consider, don't know if its worth caring about maybe Matt can comment:

> but if you are initialising the library often, it may pay to also include the zone url, otherwise it needs to be discovered by performing an API lookup

atws seems to indicate that connecting multiple times will increase the overhead of syncing, so maybe we want to make the connection once for a 'sync session', or provide a zone like they suggest.

Reply:
Also I think naming it just at_api might be good. I feel the 'object' at the end is redundant here. Mabye client instead of object?

Reply:
Yeah object is not necessary. I changed it to client. 

I'm not really sure about the best for the sync overhead point. I think its probably something that might be better to discuss later. But I've added env settings to set the api version and url to use (providing a url skips the zone api request).